plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+    // to access all graphs with component ID i, get the iterators by:
+    // auto start = std::begin(scc_component.component) + scc_component.range[i];
+    // auto end = std::begin(scc_component.component) + scc_component.range[i+1];
+    struct SCC_Component{
+        // in_component: all NodeIDs sorted by component ID
+        // in_range: index where a new component starts
+        //
+        // example: NodeID 0, 1, 2, 4, 5 are in component 0
+        //          NodeID 3, 6, 7, 8    are in component 1
+        //          => in_component = [0, 1, 2, 4, 5, 3, 6, 7, 8]
+        //          => in_range = [0, 5]
+        SCC_Component(std::vector<NodeID> in_component,
+                      std::vector<size_t> in_range)
+                      : component(in_component),
+                        range(in_range) {
+            range.push_back(in_component.size());
+            BOOST_ASSERT_MSG(in_component.size() >= in_range.size(),
+                             "scc component and its ranges do not match");
+            BOOST_ASSERT_MSG(*std::max_element(in_range.begin(), in_range.end()) < in_component.size(),
+                             "scc component ranges are out of bound");
+            BOOST_ASSERT_MSG(*std::min_element(in_range.begin(), in_range.end()) >= 0,
+                             "invalid scc component range");
+            BOOST_ASSERT_MSG([&in_range](){
+                                 for (std::size_t r = 0; r < in_range.size() - 1; ++r) {
+                                     if (in_range[r] > in_range[r+1]) {
+                                         return false;
+                                     }
+                                 }
+                                 return true;
+                             }(),
+                             "invalid component ranges");
+          };
+
+        // constructor to use when whole graph is one single scc
+        SCC_Component(std::vector<NodeID> in_component)
+                      : component(in_component), range({0, in_component.size()}) {
+                      };
+
+        std::size_t GetNumberOfComponents() const{
+            return range.size() - 1;
+        }
+
+        const std::vector<NodeID> component;
+        // component range = in_range + [component.size()]
+        std::vector<size_t> range;
+    };
+
+    // takes the number of locations and its distance matrix,
+    // identifies and splits the graph in its strongly connected components (scc)
+    // and returns an SCC_Component
+    SCC_Component SplitUnaccessibleLocations(const std::size_t number_of_locations,
+                                             const DistTableWrapper<EdgeWeight> & result_table) {
+
+        // Run TarjanSCC
+        auto wrapper = std::make_shared<MatrixGraphWrapper<EdgeWeight>>(result_table.GetTable(), number_of_locations);
+        auto scc = TarjanSCC<MatrixGraphWrapper<EdgeWeight>>(wrapper);
+        scc.run();
```

Comment:
Currently iterating over components can be done manually using [these member functions](https://github.com/Project-OSRM/osrm-backend/blob/300d901618dd259cfba408d0e3d96aeec4f40369/algorithms/tiny_components.hpp#L227-L236) from `TarjanSCC`.

If you check your code that follows, it does exactly that. I think you could move the manual iteration out to `TarjanSCC`, wrap it once with iterators, and then use it here, Because I guess iterating over components is a reasonable feature for `TarjanSCC` that is not only needed here.